### PR TITLE
parser: Fix regression in parsing of timestamp as identifier

### PR DESCRIPTION
**Pull request by moulimukherjee**
### Motivation
Fixes https://github.com/MaterializeInc/materialize/issues/22397

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - 

cc @def- In case you have more tests in mind

### CODE REVIEW

**Review by def- — APPROVED** [review]
Sorry for causing this issue with my comment.

I think the proper way of preventing this is something like Snowtrail, which we unfortunately don't have yet: https://www.snowflake.com/en/resources/report/snowtrail-testing-with-production-series-on-a-cloud-database/

**Comment by moulimukherjee** [comment]
@benesch I tried a bunch of options
- using the MaybeParseState as mentioned in https://github.com/MaterializeInc/materialize/issues/22397
- having a `parse_data_type_opt` returning a `Result<<Option<T>>, _> instead of `Result<T, _>
- and some combinations 

The issue I found was, we were correctly ignoring parse_data_type errors in some cases and since this function was getting called from multiple places, this was totally dependent upon the caller how they chose to interpret. So, there was no clean way to say something was not found vs a parsing error. In an ideal world all the parsing errors would be legitimate, but here it seems we are more like trying multiple branches of code and seeing what sticks. 

Eventually I localized the changes under the existing timestamp case, to keep the blast radius small and prioritizing the data parsing error if all parsing attempts fail.

File: src/sql-parser/src/parser.rs [hunk 1]
```diff
@@ -566,13 +566,37 @@ impl<'a> Parser<'a> {
...
+                        self.index = index;
+                        match self.parse_qualified_identifier(id.into()) {
+                            // It cannot be a function, it's a type with modifiers
+                            // Returning the original parsing error
+                            Ok(Expr::Function(_)) => Err(e),
+                            Ok(expr) => Ok(expr),
+                            Err(_) => Err(e),
+                        }
+                    }
+                }
```
Comment (benesch) [inline src/sql-parser/src/parser.rs:599]: Ugh, I see what you mean about it being hard to adapt `maybe_parse` for this. 

However, this logic is suuuuper subtle. I'm having trouble reasoning about it. It seems like a lot of complexity just to improve that one error message.

What do you think about doing a more targeted fix for this release that just removes this entire special case (and intentionally regresses the error message), and then we can experiment with some alternative methods of recovering the good error message without the time pressure.

Thanks so much for working on this one!

Comment (ParkMyCar) [reply src/sql-parser/src/parser.rs:599]: > What do you think about doing a more targeted fix for this release that just removes this entire special case (and intentionally regresses the error message), and then we can experiment with some alternative methods of recovering the good error message without the time pressure.

Personally I would also prefer this, thank you @moulimukherjee for diving deep here!

Comment (moulimukherjee) [reply src/sql-parser/src/parser.rs:599]: Sounds good 👍

**Review by benesch — APPROVED** [review]
LGTM! Thank you for slogging on this one.

**Pull request merged**